### PR TITLE
Prints message while kubectl run to reduce confusion

### PR DESCRIPTION
**Pull request by janetkuo**
Addresses kubernetes/kubernetes#23604

Add additional message to "kubectl run" default output; add a flag "--hide-message" so that users can hide it if they want to. 

``` console
$ kubectl run nginx --image=nginx
deployment "nginx" created

You have generated a deployments resource. You can feed the generated resource to a subsequent command, such as
"kubectl get deployments/nginx", or "kubectl expose deployments/nginx".
If you want to write a script to do that automatically, use "-o name" flag while doing "kubectl run", and it
will generate short output "deployments/nginx" instead, which is perfect for script input.
"--generator" flag can be used with "kubectl run" to generate other types of resources, for example, set it to
"run/v1" creates a Replication Controller, and "run-pod/v1" creates a Pod.
If you wish to hide this message when running "kubectl run", set "--hide-message=true".

$ kubectl run nginx --image=nginx -o name 
deployment/nginx

$ kubectl run nginx --image=nginx --hide-message
deployment "nginx" created
```

@kubernetes/sig-config @kubernetes/kubectl

### CODE REVIEW

**Comment by ncdc** [comment]
cc @kubernetes/rh-ux

File: pkg/kubectl/cmd/run.go [hunk 1]
```diff
@@ -274,6 +275,23 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
 		return f.PrintObject(cmd, mapper, obj, cmdOut)
 	}
 	cmdutil.PrintSuccess(mapper, false, cmdOut, mapping.Resource, args[0], "created")
+
+	// Prints 1.2 changes warning
+	if outputFormat == "" && !cmdutil.GetFlagBool(cmd, "hide-message") {
```
Comment (liggitt) [inline pkg/kubectl/cmd/run.go]: output format check is unnecessary, we already returned above if outputting

**Comment by kelseyhightower** [comment]
As an existing user of Kubernetes I would hope there was a env var I can set that hides messages like this by default. The need to type `--hide-message=true` each time does not feel right.

Is it time for a kubectl config file (.kubectlcfg)?

File: pkg/kubectl/cmd/run.go [hunk 2]
```diff
@@ -274,6 +275,23 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
...
+		generatedResourceOutput := fmt.Sprintf("%s/%s", mapping.Resource, args[0])
+		msg := fmt.Sprintf(
+			`
+You have generated a %s resource. You can feed the generated resource to a subsequent command, such as
+"kubectl get %s", or "kubectl expose %s".
+If you want to write a script to do that automatically, use "-o name" flag while doing "kubectl run", and it
+will generate short output "%s" instead, which is perfect for script input.
+"--generator" flag can be used with "kubectl run" to generate other types of resources, for example, set it to
+"run/v1" creates a Replication Controller, and "run-pod/v1" creates a Pod.
+If you wish to hide this message when running "kubectl run", set "--hide-message=true".
```
Comment (liggitt) [inline pkg/kubectl/cmd/run.go]: this seems really verbose. embedding lots of different command names in output also makes it hard to compose this command into other commands

**Comment by liggitt** [comment]
this really seems like a usage or help message, not something to include in every run

**Comment by bgrant0607** [comment]
cc @smarterclayton

**Comment by bgrant0607** [comment]
@kelseyhightower Preferences: #10693. Would love someone to work on that.

**Comment by bgrant0607** [comment]
@liggitt Yes, there is concern that the message will cause more confusion. 

We're also working on an email to kubernetes-announce and an update to our Deployment blog post.

**Comment by bgrant0607** [comment]
cc @david-mcmahon re. 1.2.1

**Comment by bgrant0607** [comment]
It would be cool if we had a way to print this just once.

**Comment by bgrant0607** [comment]
We urgently need to build 1.2.1, so let's think on this more, and work on communication for now.

cc @thockin

File: pkg/kubectl/cmd/run.go [hunk 3]
```diff
@@ -274,6 +275,23 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
 		return f.PrintObject(cmd, mapper, obj, cmdOut)
 	}
 	cmdutil.PrintSuccess(mapper, false, cmdOut, mapping.Resource, args[0], "created")
+
+	// Prints 1.2 changes warning
+	if outputFormat == "" && !cmdutil.GetFlagBool(cmd, "hide-message") {
+		generatedResourceOutput := fmt.Sprintf("%s/%s", mapping.Resource, args[0])
+		msg := fmt.Sprintf(
+			`
```
Comment (thockin) [inline pkg/kubectl/cmd/run.go]: This would be a good place to use `github.com/renstrom/dedent`

File: pkg/kubectl/cmd/run.go [hunk 4]
```diff
@@ -274,6 +275,23 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
...
+You have generated a %s resource. You can feed the generated resource to a subsequent command, such as
+"kubectl get %s", or "kubectl expose %s".
+If you want to write a script to do that automatically, use "-o name" flag while doing "kubectl run", and it
```
Comment (thockin) [inline pkg/kubectl/cmd/run.go]: use the "-o name" flag - the "the" makes it read better

Comment (thockin) [inline pkg/kubectl/cmd/run.go]: s/while doing/to/

File: pkg/kubectl/cmd/run.go [hunk 5]
```diff
@@ -274,6 +275,23 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
...
+will generate short output "%s" instead, which is perfect for script input.
+"--generator" flag can be used with "kubectl run" to generate other types of resources, for example, set it to
```
Comment (thockin) [inline pkg/kubectl/cmd/run.go]: The "--generator" - the "the" makes it easier to read.

Comment (thockin) [inline pkg/kubectl/cmd/run.go]: types of resources.  For example: set it

Comment (thockin) [inline pkg/kubectl/cmd/run.go]: setting it

File: pkg/kubectl/cmd/run.go [hunk 6]
```diff
@@ -274,6 +275,23 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
...
+"run/v1" creates a Replication Controller, and "run-pod/v1" creates a Pod.
```
Comment (thockin) [inline pkg/kubectl/cmd/run.go]: s/Replication Controller/ReplicationController/

File: pkg/kubectl/cmd/run.go (hunk 2 above)
Comment (thockin) [inline pkg/kubectl/cmd/run.go]: To hide this message set the "--hide-message" flag.

**Comment by thockin** [comment]
I like kelsey's idea of a `~/.kubectlrc` file, but that's a whole different design doc.  I could see an env var to mute this.  Maybe `K8S_STIFLE_FOLLOWUP_MSGS` (any non-empty value means true) - it could also apply to the `expose` message about firewalls.

I suggested in another bug that we denote these messages syntactically with a leading `#`.  I know the default output is for humans, but I still think it would read better, and it would allow you to use blank lines between parapgraphs.

```
# You have generated a deployments resource. You can feed the generated resource to a subsequent command, such as
# "kubectl get deployments/nginx", or "kubectl expose deployments/nginx".
#
# If you want to write a script to do that automatically, use "-o name" flag while doing "kubectl run", and it
# will generate short output "deployments/nginx" instead, which is perfect for script input.
#
# The "--generator" flag can be used with "kubectl run" to generate other types of resources, for example, set it to
#"run/v1" creates a Replication Controller, and "run-pod/v1" creates a Pod.
#
# If you wish to hide this message when running "kubectl run", set "--hide-message=true".
```

**Comment by smarterclayton** [comment]
At some point we probably want to just accept that the "run" command
is a user interface, and use a judicious amount of space to
communicate what we are doing.  We have "-o" specifically to handle
scripting or sophisticated users, but bare run is for humans.

What matters for end users on run?

1. What is happening (if anything is opaque, or if run is violating
their expectation)
2. What is created (if the things being generated do not match expectations)
3. What to do next (if the outcome is not clear)

Pretty sure instead of worrying about suppression, we should
succinctly and clearly communicate the above by default (I agree with
Jordan are telling the user far too much)


"Creating a job "foo" to run this image, use kubectl describe job/foo
to see details of execution.  See kubectl help run for other options."

We should probably only print the second sentence if (as noted above)
the user did a simple run (where they could be expected to not know
what to do)

kubectl help run should be better at showing users what they actually
need to know.  We need to elide useless options, have better
description, and probably switch to enabling the system pager.  That's
a good portion of the problem.

**Comment by bgrant0607** [comment]
kubectl suggesting followup commands: #19736

**Comment by bgrant0607** [comment]
Closing this for now.

**Pull request closed without merging**